rust/kernel/net/ethtool.rs

```diff
@@ -0,0 +1,252 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Ethtool Operations.
+
+use core::marker;
+
+use crate::error::{Error, Result};
+use crate::types::{SavedAsPointer, SavedAsPointerMut};
+use crate::{bindings, c_types, from_kernel_result};
+
+use super::device::{NetDevice, NetDeviceAdapter};
+
+unsafe extern "C" fn get_drvinfo_callback<T: NetDeviceAdapter>(
+    dev: *mut bindings::net_device,
+    info: *mut bindings::ethtool_drvinfo,
+) {
+    T::EthOps::get_drvinfo(
+        unsafe { &mut NetDevice::<T>::from_pointer_mut(dev) },
+        unsafe { &mut EthtoolDrvinfo::from_pointer(info) },
+    );
+}
+
+unsafe extern "C" fn get_ts_info_callback<T: NetDeviceAdapter>(
+    dev: *mut bindings::net_device,
+    info: *mut bindings::ethtool_ts_info,
+) -> c_types::c_int {
+    from_kernel_result! {
+        T::EthOps::get_ts_info(
+            // SAFETY: dev is valid, as this is a callback
+            unsafe { &mut NetDevice::<T>::from_pointer_mut(dev) },
+            // SAFETY: info is valid, as this is a callback
+            unsafe { &mut EthToolTsInfo::from_pointer(info) }
+        )?;
+        Ok(0)
+    }
+}
+
+pub(crate) struct EthToolOperationsVtable<T: NetDeviceAdapter>(marker::PhantomData<T>);
+
+impl<T: NetDeviceAdapter> EthToolOperationsVtable<T> {
+    const VTABLE: bindings::ethtool_ops = bindings::ethtool_ops {
+        #[cfg(CONFIG_X86)]
+        _bitfield_align_1: [],
+
+        #[cfg(CONFIG_X86)]
+        _bitfield_1: bindings::__BindgenBitfieldUnit::<[u8; 1usize]>::new([0u8; 1usize]),
+
+        #[cfg(not(CONFIG_x86))]
+        _bitfield_1: bindings::__BindgenBitfieldUnit::<[u8; 1usize], u8>::new([0u8; 1usize]),
```

Comment:
This bit of code caused compilation issues on x86 (and for other platforms).

1. For `_bitfield_1`:

    ```
    error[E0107]: this struct takes 2 generic arguments but 1 generic argument was supplied
      --> rust/kernel/net/ethtool.rs:46:32
       |
    46 |         _bitfield_1: bindings::__BindgenBitfieldUnit::<[u8; 1usize]>::new([0u8; 1usize]),
       |                                ^^^^^^^^^^^^^^^^^^^^^   ------------ supplied 1 generic argument
       |                                |
       |                                expected 2 generic arguments
       |
    note: struct defined here, with 2 generic parameters: `Storage`, `Align`
      --> /home/ds/code/linux/rust/bindings_generated.rs:5:12
       |
    5  | pub struct __BindgenBitfieldUnit<Storage, Align> {
       |            ^^^^^^^^^^^^^^^^^^^^^ -------  -----
    help: add missing generic argument
       |
    46 |         _bitfield_1: bindings::__BindgenBitfieldUnit::<[u8; 1usize], Align>::new([0u8; 1usize]),
       |
    ```

   Basically the `#[cfg(not(CONFIG_X86)]` variant of this field could be used for every platform.

2. `_bitfield_align_1` field is not generated by bindgen (I have tested with version `bindgen 0.56.0`).

Reply:
Interesting that it got generated on my side. Although only did brief testing on x86, could have been an outdated bindgen.

Reply:
Removed the cfg, will wait for all green check until resolving, but looks good so far, do not have a x86 right now to test, so will have to wait for CI.